File: Análise e Complexidade de Algoritmos/radixSort/radix.py
import random
import time

def digito(num, dig):
    for _ in range(dig):
        num = int(num / 10)
    return num % 10


def countingSortEst(v, d):
    count = [0] * 10
    temp = list(v)
    for el in range(len(v)):
        count[digito(el, d)] += 1

    for i in range(1, len(count)):
        count[i] += count[i - 1]

    for el in reversed(v):
        temp[count[digito(el, d)] - 1] = el
        count[digito(el, d)] -= 1
    return temp

def radix(v):
  d = 10
  for i in range(d):
    v = countingSortEst(v, i)
  return v


v = [123, 434, 255, 656, 744, 888]


for i in range(10):
  n =i*100
  vetor = [random.randint(0,9999) for _ in range(n)]
  inicio = time.process_time_ns()
  radix(vetor)
  dt = time.process_time_ns() - inicio 
  print(n, dt)

File: Análise e Complexidade de Algoritmos/radixSort/radix.cpp
#include <iostream>
#include <vector>

using namespace std;

int digito(int num, int dig){
  for (int i= 0; i < dig; i++){
    num /= 10;
  }
  return num % 10;
}

void countingSort(vector<int> &v, int d) {
    vector<int> counting(10,0);
    vector<int> temp = v;

  for(auto &e : v){
    counting[digito(e, d)]++;
  }

  for(int i = 1; i < counting.size(); i++){
    counting[i] +=counting[i -1];
  }

  for(int i = v.size() -1; i >=0; i--){
    temp[counting[digito(v[i], d)] -1] = v[i];
    counting[digito(v[i], d)] -=1;
  }

  v = temp;
}

int main() {
  vector<int> v={532,243,112,420,423,999,116};

  countingSort(v, 0);
  countingSort(v, 1);
  countingSort(v, 2);

  for(auto& e: v){
    cout << e << " ";
  }

  cout << endl; 
  return 0;
}